File: packaging/beam-languages/elixir/lib/sshfling.ex
defmodule SSHFling do
  @moduledoc "Launcher API for the bundled SSHFling runtime."

  @spec runtime_path() :: String.t()
  def runtime_path do
    configured_or("SSHFLING_RUNTIME", Application.app_dir(:sshfling, "priv/runtime/sshfling.py"))
  end

  @spec template_directory() :: String.t()
  def template_directory do
    configured_or("SSHFLING_TEMPLATE_DIR", Application.app_dir(:sshfling, "priv/runtime/templates"))
  end

  @spec run([String.t()]) :: non_neg_integer()
  def run(arguments) when is_list(arguments) do
    if Enum.all?(arguments, &is_binary/1) do
      python = configured_or("SSHFLING_PYTHON", "python3")
      runtime = runtime_path()

      if not File.regular?(runtime) do
        127
      else
        try do
        {_output, status} =
          System.cmd(python, [runtime | arguments],
            env: [
              {"SSHFLING_TEMPLATE_DIR", template_directory()},
              {"PYTHONUNBUFFERED", "1"}
            ],
            into: IO.stream(:stdio, :line),
            stderr_to_stdout: true
          )

          status
        rescue
          error in ErlangError ->
            IO.puts(:stderr, "sshfling: #{Exception.message(error)}")
            127
        end
      end
    else
      raise ArgumentError, "SSHFling arguments must all be strings"
    end
  end

  defp configured_or(name, default) do
    case System.get_env(name) do
      nil -> default
      "" -> default
      value -> value
    end
  end
end

File: packaging/beam-languages/erlang/src/sshfling.erl
-module(sshfling).

-export([run/1, runtime_path/0, template_directory/0]).

-spec environment_or_default(string(), string()) -> string().
environment_or_default(Name, Default) ->
    case os:getenv(Name) of
        false -> Default;
        "" -> Default;
        Value -> Value
    end.

-spec package_root() -> file:filename_all().
package_root() ->
    case os:getenv("SSHFLING_PACKAGE_ROOT") of
        false ->
            case code:priv_dir(sshfling) of
                {error, bad_name} -> filename:absname(".");
                Priv -> filename:dirname(Priv)
            end;
        "" -> filename:absname(".");
        Root -> Root
    end.

-spec runtime_path() -> file:filename_all().
runtime_path() ->
    environment_or_default(
      "SSHFLING_RUNTIME",
      filename:join([package_root(), "priv", "runtime", "sshfling.py"])).

-spec template_directory() -> file:filename_all().
template_directory() ->
    environment_or_default(
      "SSHFLING_TEMPLATE_DIR",
      filename:join([package_root(), "priv", "runtime", "templates"])).

-spec executable(string()) -> string() | false.
executable(Python) ->
    case filename:pathtype(Python) of
        absolute -> Python;
        _ -> os:find_executable(Python)
    end.

-spec run([string()]) -> non_neg_integer().
run(Arguments) when is_list(Arguments) ->
    true = lists:all(fun is_list/1, Arguments),
    Python = environment_or_default("SSHFLING_PYTHON", "python3"),
    Runtime = runtime_path(),
    case {filelib:is_regular(Runtime), executable(Python)} of
        {false, _} -> 127;
        {_, false} -> 127;
        {true, Program} ->
            Port = open_port(
                     {spawn_executable, Program},
                     [binary, exit_status, use_stdio, stderr_to_stdout,
                      {args, [Runtime | Arguments]},
                      {env, [{"SSHFLING_TEMPLATE_DIR", template_directory()},
                             {"PYTHONUNBUFFERED", "1"}]}]),
            wait_for_exit(Port)
    end.

-spec wait_for_exit(port()) -> non_neg_integer().
wait_for_exit(Port) ->
    receive
        {Port, {data, Data}} ->
            ok = io:put_chars(Data),
            wait_for_exit(Port);
        {Port, {exit_status, Status}} ->
            Status
    end.
